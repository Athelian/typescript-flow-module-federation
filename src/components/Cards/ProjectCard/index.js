// @flow
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { isForbidden } from 'utils/data';
import { Display, FieldItem, Label } from 'components/Form';
import FormattedDate from 'components/FormattedDate';
import FormattedNumber from 'components/FormattedNumber';
import Divider from 'components/Divider';
import TaskStatusChart from 'components/TaskStatusChart';
import withForbiddenCard from 'hoc/withForbiddenCard';
import Tag from 'components/Tag';
import BaseCard from '../BaseCard';
import {
  ProjectCardStyle,
  ProjectNameStyle,
  CommonCardGridStyle,
  TaskStatusChartStyle,
  TagsWrapperStyle,
} from './style';

type OptionalProps = {
  onClick: Function,
};

type Props = OptionalProps & {
  project: {
    name: string,
    dueDate: string,
    milestones: Array<Object>,
    taskCount: {
      count: number,
      remain: number,
      inProgress: number,
      completed: number,
      delayed: number,
    },
    tags: Array<Object>,
  },
};

const ProjectCard = ({ project, onClick }: Props) => {
  const { name, dueDate, milestones = [], taskCount, tags = [] } = project;
  const { count, remain, inProgress, completed, delayed } = taskCount;

  return (
    <BaseCard icon="PROJECT" color="PROJECT" onClick={onClick}>
      <div className={ProjectCardStyle}>
        <div className={CommonCardGridStyle}>
          <div className={ProjectNameStyle}>
            <Display align="left">{name}</Display>
          </div>

          <FieldItem
            label={
              <Label>
                <FormattedMessage id="components.cards.dueDate" defaultMessage="DUE DATE" />
              </Label>
            }
            input={
              <Display>
                <FormattedDate value={dueDate} />
              </Display>
            }
          />

          <FieldItem
            label={
              <Label>
                <FormattedMessage id="components.cards.milestones" defaultMessage="MILESTONES" />
              </Label>
            }
            input={
              <Display>
                <FormattedNumber value={milestones.length} />
              </Display>
            }
          />

          <Divider />

          <FieldItem
            label={
              <Label>
                <FormattedMessage id="components.cards.tasks" defaultMessage="TASKS" />
              </Label>
            }
            input={
              <Display>
                <FormattedNumber value={count} />
              </Display>
            }
          />

          <FieldItem
            label={
              <Label>
                <FormattedMessage id="components.cards.overdueTasks" defaultMessage="OVERDUE" />
              </Label>
            }
            input={
              <Display>
                <FormattedNumber value={delayed} />
              </Display>
            }
          />

          <div className={TaskStatusChartStyle}>
            <TaskStatusChart
              completed={completed}
              inProgress={inProgress}
              // FIXME: wait for API, ref: https://zenport.slack.com/archives/C2JTDSRJ6/p1560920001006800
              skipped={0}
              unCompleted={remain}
            />
          </div>

          <div className={TagsWrapperStyle}>
            {tags
              .filter(item => !isForbidden(item))
              .map(tag => (
                <Tag key={tag.id} tag={tag} />
              ))}
          </div>
        </div>
      </div>
    </BaseCard>
  );
};

export default withForbiddenCard(ProjectCard, 'project', {
  width: '195px',
  height: '214px',
  entityIcon: 'PROJECT',
  entityColor: 'PROJECT',
});
