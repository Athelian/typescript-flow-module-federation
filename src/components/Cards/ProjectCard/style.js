// @flow
import { css } from 'react-emotion';

import { layout } from 'styles/common';

export const ProjectCardStyle: string = css`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 195px;
  height: 242px;
`;

export const CommonCardGridStyle: string = css`
  display: grid;
  grid-template-columns: 185px;
  grid-gap: 5px;
  padding: 5px 5px 10px 5px;
`;

export const ProjectNameStyle: string = css`
  width: 100%;
  padding: 0 15px 0 0;
`;

export const TagsWrapperStyle: string = css`
  ${layout.GRID_HORIZONTAL};
  grid-gap: 5px;
  overflow: hidden;
  width: 100%;
`;
