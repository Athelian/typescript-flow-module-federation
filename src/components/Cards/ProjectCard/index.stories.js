import React from 'react';
/* eslint-disable import/no-extraneous-dependencies */
import { storiesOf } from '@storybook/react';
import { IntlProvider } from 'react-intl';
import ProjectCard from './index';

storiesOf('Card', module).add('Project Card', () => (
  <IntlProvider>
    <div
      style={{
        display: 'grid',
        gridGap: 5,
      }}
    >
      <div>
        <ProjectCard
          project={{
            name: 'PROJECT TITLE',
            dueDate: '2019-6-30',
            taskCount: {
              count: 20,
              remain: 1,
              inProgress: 2,
              completed: 2,
              rejected: 2,
              approved: 2,
              delayed: 2,
            },
            tags: [
              { id: 1, name: 'tag1', color: '#123456' },
              { id: 2, name: 'tag2', color: '#FF00FF' },
              { id: 3, name: 'tag2', color: '#FF00FF' },
              { id: 4, name: 'tag2', color: '#FF00FF' },
              { id: 5, name: 'tag2', color: '#FF00FF' },
            ],
          }}
        />
      </div>

      <div>
        <ProjectCard
          project={{
            name: 'PROJECT long long long long title',
            dueDate: '2019-6-30',
            taskCount: {
              taskCount: {
                count: 20,
                remain: 1,
                inProgress: 2,
                completed: 2,
                delayed: 2,
              },
            },
            tags: [],
          }}
        />
      </div>
    </div>
  </IntlProvider>
));
