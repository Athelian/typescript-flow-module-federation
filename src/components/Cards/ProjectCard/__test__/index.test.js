import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import ProjectCard from '../index';

describe('Cards', () => {
  test('render ProjectCard', () => {
    const project = {
      name: 'PROJECT TITLE',
      dueDate: '2019-6-30',
      taskCount: {
        count: 20,
        remain: 1,
        inProgress: 2,
        completed: 2,
        rejected: 2,
        approved: 2,
        delayed: 2,
      },
      tags: [
        { id: 1, name: 'tag1', color: '#123456' },
        { id: 2, name: 'tag2', color: '#FF00FF' },
        { id: 3, name: 'tag2', color: '#FF00FF' },
        { id: 4, name: 'tag2', color: '#FF00FF' },
        { id: 5, name: 'tag2', color: '#FF00FF' },
      ],
    };
    const { container, asFragment } = render(
      <IntlProvider locale="en">
        <ProjectCard project={project} />
      </IntlProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(asFragment()).toMatchSnapshot();
  });
});
