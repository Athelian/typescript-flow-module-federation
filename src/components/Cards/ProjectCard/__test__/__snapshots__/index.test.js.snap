// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards render ProjectCard 1`] = `
<div
  class="css-16dyits"
  id=""
>
  <button
    class="css-1tmldnl"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-th-list fa-w-16 fa-fw fa-rotate-90 "
      data-icon="th-list"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M149.333 216v80c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24v-80c0-13.255 10.745-24 24-24h101.333c13.255 0 24 10.745 24 24zM0 376v80c0 13.255 10.745 24 24 24h101.333c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H24c-13.255 0-24 10.745-24 24zM125.333 32H24C10.745 32 0 42.745 0 56v80c0 13.255 10.745 24 24 24h101.333c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24zm80 448H488c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24zm-24-424v80c0 13.255 10.745 24 24 24H488c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24zm24 264H488c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24z"
        fill="currentColor"
      />
    </svg>
  </button>
  <div
    class="css-1wsm69r"
  >
    <div
      class="css-5vwma5"
    >
      <div
        class="css-1ymjl0v"
      >
        <div
          class="css-19flw6m"
        >
          PROJECT TITLE
        </div>
      </div>
      <div
        class="css-19wgwnm"
      >
        <div
          class="css-12hroh9"
        >
          <span>
            DUE DATE
          </span>
        </div>
        <div
          class="css-b6ifgt"
        >
          <span>
            6/30/2019
          </span>
        </div>
      </div>
      <div
        class="css-19wgwnm"
      >
        <div
          class="css-12hroh9"
        >
          <span>
            MILESTONES
          </span>
        </div>
        <div
          class="css-b6ifgt"
        >
          
          <span>
            0
          </span>
          
        </div>
      </div>
      <div
        class="css-ar71m2"
      />
      <div
        class="css-19wgwnm"
      >
        <div
          class="css-12hroh9"
        >
          <span>
            TASKS
          </span>
        </div>
        <div
          class="css-b6ifgt"
        >
          
          <span>
            20
          </span>
          
        </div>
      </div>
      <div
        class="css-19wgwnm"
      >
        <div
          class="css-12hroh9"
        >
          <span>
            OVERDUE
          </span>
        </div>
        <div
          class="css-b6ifgt"
        >
          
          <span>
            2
          </span>
          
        </div>
      </div>
      <div
        class="css-ar7jq1"
      >
        <div
          class="css-10q8dqb"
        >
          <div
            class="css-toukby"
          >
            <div
              class="css-yedefi"
            />
            <div
              class="css-15i5amq"
            />
            <div
              class="css-14pjm7c"
            />
          </div>
          <div
            class="css-1lwluhx"
          >
            <div
              class="css-7d1hvd"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-check-circle fa-w-16 fa-fw "
                data-icon="check-circle"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                  fill="currentColor"
                />
              </svg>
              2
            </div>
            <div
              class="css-29mzby"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-stopwatch fa-w-14 fa-fw "
                data-icon="stopwatch"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432 304c0 114.9-93.1 208-208 208S16 418.9 16 304c0-104 76.3-190.2 176-205.5V64h-28c-6.6 0-12-5.4-12-12V12c0-6.6 5.4-12 12-12h120c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-28v34.5c37.5 5.8 71.7 21.6 99.7 44.6l27.5-27.5c4.7-4.7 12.3-4.7 17 0l28.3 28.3c4.7 4.7 4.7 12.3 0 17l-29.4 29.4-.6.6C419.7 223.3 432 262.2 432 304zm-176 36V188.5c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12V340c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12z"
                  fill="currentColor"
                />
              </svg>
              2
            </div>
            <div
              class="css-ricg38"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-forward fa-w-16 fa-fw "
                data-icon="forward"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M500.5 231.4l-192-160C287.9 54.3 256 68.6 256 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2zm-256 0l-192-160C31.9 54.3 0 68.6 0 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2z"
                  fill="currentColor"
                />
              </svg>
              0
            </div>
            <div
              class="css-72tgzx"
            >
              1
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-udh7f1"
      >
        <div
          class="css-1gjilff"
        >
          tag1
        </div>
        <div
          class="css-1sdky1k"
        >
          tag2
        </div>
        <div
          class="css-1sdky1k"
        >
          tag2
        </div>
        <div
          class="css-1sdky1k"
        >
          tag2
        </div>
        <div
          class="css-1sdky1k"
        >
          tag2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Cards render ProjectCard 2`] = `
<DocumentFragment>
  <div
    class="css-16dyits"
    id=""
  >
    <button
      class="css-1tmldnl"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-th-list fa-w-16 fa-fw fa-rotate-90 "
        data-icon="th-list"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M149.333 216v80c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24v-80c0-13.255 10.745-24 24-24h101.333c13.255 0 24 10.745 24 24zM0 376v80c0 13.255 10.745 24 24 24h101.333c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H24c-13.255 0-24 10.745-24 24zM125.333 32H24C10.745 32 0 42.745 0 56v80c0 13.255 10.745 24 24 24h101.333c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24zm80 448H488c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24zm-24-424v80c0 13.255 10.745 24 24 24H488c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24zm24 264H488c13.255 0 24-10.745 24-24v-80c0-13.255-10.745-24-24-24H205.333c-13.255 0-24 10.745-24 24v80c0 13.255 10.745 24 24 24z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="css-1wsm69r"
    >
      <div
        class="css-5vwma5"
      >
        <div
          class="css-1ymjl0v"
        >
          <div
            class="css-19flw6m"
          >
            PROJECT TITLE
          </div>
        </div>
        <div
          class="css-19wgwnm"
        >
          <div
            class="css-12hroh9"
          >
            <span>
              DUE DATE
            </span>
          </div>
          <div
            class="css-b6ifgt"
          >
            <span>
              6/30/2019
            </span>
          </div>
        </div>
        <div
          class="css-19wgwnm"
        >
          <div
            class="css-12hroh9"
          >
            <span>
              MILESTONES
            </span>
          </div>
          <div
            class="css-b6ifgt"
          >
            <span>
              0
            </span>
          </div>
        </div>
        <div
          class="css-ar71m2"
        />
        <div
          class="css-19wgwnm"
        >
          <div
            class="css-12hroh9"
          >
            <span>
              TASKS
            </span>
          </div>
          <div
            class="css-b6ifgt"
          >
            <span>
              20
            </span>
          </div>
        </div>
        <div
          class="css-19wgwnm"
        >
          <div
            class="css-12hroh9"
          >
            <span>
              OVERDUE
            </span>
          </div>
          <div
            class="css-b6ifgt"
          >
            <span>
              2
            </span>
          </div>
        </div>
        <div
          class="css-ar7jq1"
        >
          <div
            class="css-10q8dqb"
          >
            <div
              class="css-toukby"
            >
              <div
                class="css-yedefi"
              />
              <div
                class="css-15i5amq"
              />
              <div
                class="css-14pjm7c"
              />
            </div>
            <div
              class="css-1lwluhx"
            >
              <div
                class="css-7d1hvd"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-check-circle fa-w-16 fa-fw "
                  data-icon="check-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                    fill="currentColor"
                  />
                </svg>
                2
              </div>
              <div
                class="css-29mzby"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-stopwatch fa-w-14 fa-fw "
                  data-icon="stopwatch"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M432 304c0 114.9-93.1 208-208 208S16 418.9 16 304c0-104 76.3-190.2 176-205.5V64h-28c-6.6 0-12-5.4-12-12V12c0-6.6 5.4-12 12-12h120c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-28v34.5c37.5 5.8 71.7 21.6 99.7 44.6l27.5-27.5c4.7-4.7 12.3-4.7 17 0l28.3 28.3c4.7 4.7 4.7 12.3 0 17l-29.4 29.4-.6.6C419.7 223.3 432 262.2 432 304zm-176 36V188.5c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12V340c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12z"
                    fill="currentColor"
                  />
                </svg>
                2
              </div>
              <div
                class="css-ricg38"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-forward fa-w-16 fa-fw "
                  data-icon="forward"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M500.5 231.4l-192-160C287.9 54.3 256 68.6 256 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2zm-256 0l-192-160C31.9 54.3 0 68.6 0 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2z"
                    fill="currentColor"
                  />
                </svg>
                0
              </div>
              <div
                class="css-72tgzx"
              >
                1
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-udh7f1"
        >
          <div
            class="css-1gjilff"
          >
            tag1
          </div>
          <div
            class="css-1sdky1k"
          >
            tag2
          </div>
          <div
            class="css-1sdky1k"
          >
            tag2
          </div>
          <div
            class="css-1sdky1k"
          >
            tag2
          </div>
          <div
            class="css-1sdky1k"
          >
            tag2
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
